lib/sensors/src/tof.rs

```diff
@@ -0,0 +1,205 @@
+    i2c: T,
+    device_address: u8,
+}
+
+impl<T: HypedI2c> TimeOfFlight<T> {
+    // Create a new instance of time of flight sensor, configure
+    pub fn new(mut i2c: T, device_address: ToFAddresses) -> Result<Self, ToFError> { // SR03 Settings as seen in Application Sheet
+        let device_address = device_address as u8;
+        for i in 0..30 { // writing to private registers
+        match i2c.write_byte_to_register_16(device_address, PRIVATE_REGISTERS[i], PRIVATE_REGISTER_DATA[i]) {
+            Ok(p) => p, // unused
+            Err(e) =>  return Err(ToFError::I2cError(e)),
+        }
+    } 
+        // Recommended Public Registers now (see Application Sheet)
+        match i2c.write_byte_to_register(device_address, SYS_MODE_GPIO1, SYS_MODE_GPIO_VAL) {
+            Ok(p) => p,
+            Err(e) =>  return Err(ToFError::I2cError(e)),
+        }
+        match i2c.write_byte_to_register_16(device_address, AVG_SAMPLE_PERIOD, AVG_SAMPLE_PERIOD_VAL) {
+            Ok(p) => p,
+            Err(e) =>  return Err(ToFError::I2cError(e)),
+        }
+        match i2c.write_byte_to_register(device_address, SYSRANGE_VHV_REPEAT_RATE, SYSRANGE_VHV_REPEAT_RATE_VAL) {
+            Ok(p) => p,
+            Err(e) =>  return Err(ToFError::I2cError(e)),
+        }
+        match i2c.write_byte_to_register(device_address, SYSRANGE_VHV_RECALIBRATE, SYSRANGE_VHV_RECALIBRATE_VAL) {
+            Ok(p) => p,
+            Err(e) =>  return Err(ToFError::I2cError(e)),
+        }
+        match i2c.write_byte_to_register(device_address, SYSRANGE_INTERMEASURE_PERIOD, SYSRANGE_INTERMEASURE_PERIOD_VAL) {
+            Ok(p) => p,
+            Err(e) =>  return Err(ToFError::I2cError(e)),
+        }
+        match i2c.write_byte_to_register(device_address, SYS_INTERRUPT_CONFIG_GPIO, SYS_INTERRUPT_CONFIG_GPIO_VAL) {
+            Ok(p) => p,
+            Err(e) =>  return Err(ToFError::I2cError(e)),
+        }
+
+        // not sure if this works? effectively if no errors returned from previous match arms, return Ok(Self)
+        Ok(Self {
+            i2c,
+            device_address,
+        })
+    }
+
+    pub fn start_ss_measure(mut i2c: T, device_address: ToFAddresses) -> Result<Self, ToFError> { // start single shot measurement
+        let device_address = device_address as u8;
+        match i2c.write_byte_to_register(device_address, SYSRANGE_START, SYSRANGE_START_SS_VAL) {
+            Ok(p) => p,
+            Err(e) =>  return Err(ToFError::I2cError(e)),
+        }
+        
+        Ok (Self {
+            i2c,
+            device_address
+        })
+    }
+
+    pub fn poll_range(&mut self) {
+        let status_byte = 
+        match self.i2c.read_byte(self.device_address, RESULT_INTERR_STATUS_GPIO) {
+            Some(byte) => byte,
+            None => 0, // not sure about returning 0 for None - will this somehow break stuff?
+        };
+        let range_status = status_byte & 0x07;
+        while (range_status != 0x04) {
```

Comment:
Yeah I'm not sure about this loop, there must be an alternative way of doing it which is less scary